### ISSUE TITLE
Logstash crashes on invalid SSL input - IOError: closed stream

### ISSUE THREAD

**Issue by jamessemai** [logstash-plugins/logstash-input-tcp#34]
(This issue was originally filed at https://github.com/elastic/logstash/issues/4366)

**Problem**: 
We noticed that, after upgrading from Logstash 1.5.5 to Logstash 2.1.1 on Linux (RHEL 6.7 and CentOS 6.7), our logstash servers would stop responding after performing a vulnerability scan on the SSL-enabled TCP input ports.

They would crash after logging the following entry to ``logstash.err``:

```
IOError: closed stream
                  peeraddr at org/jruby/ext/socket/RubyIPSocket.java:95
                  peeraddr at org/jruby/ext/openssl/SSL.java:209
                      peer at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/util/socket_peer.rb:5
             handle_socket at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:171
  server_connection_thread at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:144
```

**How to reproduce:**
After some investigation, we are able to reproduce this reliably using the following minimal commands:

*(1) Run logstash:*

```
 /opt/logstash/bin/logstash -f test.conf
```

with ``test.conf``:
```
input {
      tcp {
             codec => multiline {
                       pattern => "^<"
                       negate => true
                       what => previous
                      }
             port => 6514
             ssl_cert => "/etc/logstash/tls/logstash.crt"
             ssl_key => "/etc/logstash/tls/logstash.key"
             ssl_enable => true
             ssl_verify => false
      }
}
output {
    file {
        path => "/data/logs/test.log"
    }
}

```

Note: the logstash SSL certificate is signed by a self-signed CA.

*(2) Run following command to crash logstash:*

```
nmap -n -p6514 -sV localhost & sleep 1 && openssl s_client -connect localhost:6514
```

Note: the ``sleep 1`` ensures that the openssl command only starts after nmap has had time to start. If you do not get the "CONNECTED(00000003)" in the output below, you might want to increase this a little bit.

Output of test command (sanitizing some of the output):

```
[root@logstash]#  nmap -n -p6514 -sV localhost & sleep 1 && openssl s_client -connect localhost:6514
[1] 32603

Starting Nmap 5.51 ( http://nmap.org ) at 2015-12-17 11:27 CET
CONNECTED(00000003)
depth=0 C = *, ST = *, O = *, OU = *, CN = *, emailAddress = *
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = *, ST = *, O = *, OU = *, CN = *, emailAddress = *
verify error:num=27:certificate not trusted
verify return:1
depth=0 C = *, ST = *, O = *, OU = *, CN = *, emailAddress = *
verify error:num=21:unable to verify the first certificate
verify return:1
---
Certificate chain
[...]
---
Server certificate
-----BEGIN CERTIFICATE-----
[...]
-----END CERTIFICATE-----
[...]
---
No client certificate CA names sent
Server Temp Key: ECDH, secp521r1, 521 bits
---
SSL handshake has read 1658 bytes and written 467 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-SHA
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-SHA
    Session-ID: *
    Session-ID-ctx: 
    Master-Key: *
    Key-Arg   : None
    Krb5 Principal: None
    PSK identity: None
    PSK identity hint: None
    Start Time: 1450348041
    Timeout   : 300 (sec)
    Verify return code: 21 (unable to verify the first certificate)
---
closed
[root@logstash]#  Nmap scan report for localhost (127.0.0.1)
Host is up (0.000058s latency).
PORT     STATE SERVICE VERSION
6514/tcp open  unknown

Service detection performed. Please report any incorrect results at http://nmap.org/submit/ .
Nmap done: 1 IP address (1 host up) scanned in 11.60 seconds

[1]+  Done                    nmap -n -p6514 -sV localhost
```

Output from logstash:

```
[root@logstash]#  /opt/logstash/bin/logstash -f test.conf 
Settings: Default filter workers: 1
Logstash startup completed
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Socket closed>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error}
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Unrecognized SSL message, plaintext connection?>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error}
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Unrecognized SSL message, plaintext connection?>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error}
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Unrecognized SSL message, plaintext connection?>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error}
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Unrecognized SSL message, plaintext connection?>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error}
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Unrecognized SSL message, plaintext connection?>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error}
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Unrecognized SSL message, plaintext connection?>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error}
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Unrecognized SSL message, plaintext connection?>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error}
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Client requested protocol SSLv3 not enabled or not supported>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error}
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Unrecognized SSL message, plaintext connection?>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error}
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Unrecognized SSL message, plaintext connection?>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error}
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Unrecognized SSL message, plaintext connection?>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error}
IOError: closed stream
                  peeraddr at org/jruby/ext/socket/RubyIPSocket.java:95
                  peeraddr at org/jruby/ext/openssl/SSL.java:209
                      peer at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/util/socket_peer.rb:5
             handle_socket at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:171
  server_connection_thread at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.0/lib/logstash/inputs/tcp.rb:144
[root@logstash]# 
```

**Closing remarks:**
I am using the multiline codec as that ensured a more reliable reproduction of the issue. However, in my testing, it seems that the multiline codec is not required for the code to crash - I was also able to reproduce full logstash crashes without - just not reliably. 
Note that, in all cases without multiline code, logstash will close all existing SSL connections (but not always crash) - which is an issue in itself, but of lesser severity. I am assuming both of these issues are related, so not opening a separate ticket on that for now.

**Comment by ranman** [comment]
I can confirm that I experience this same behavior.

**Comment by pik** [comment]
For a test I replaced logstash-tcp-input with a quick python script: https://gist.github.com/pik/d9f6674b5bd20e90dc63

I was receiving input from Heroku (so maybe a dozen different peers) -- I think one of them has SSL params which are invalid (at least with the create_default_context SSL settings - these are provided by OpenSSL I think). A quick try/accept to drop the peer allows the rest of the process to run. 

Now this may seem like a minor SSL Issue and should be a quick fix - but there are multiple related tickets which appear to have been open for days and/or months without response - not to mention logstash-tcp-input crashes entirely rather than merely disconnecting a single peer - and crashes it's parent process as a result. This is very bad.

**Comment by ranman** [comment]
I'd advise everyone who reaches this comment to give up on doing TLS in logstash -- it is poorly designed and implemented (no offense to the authors -- this stuff is hard). We ended up just terminating TLS at a different process and then forwarding it on into logstash unecrypted. You can do that with various approaches (e.g. HAProxy).

**Comment by pik** [comment]
> I'd advise everyone who reaches this comment to give up on doing TLS in logstash -- it is poorly designed and implemented (no offense to the authors -- this stuff is hard).

Yes, that seems to be my impression as well - but in that case this use-case should not be encouraged on the official documentation imho. I'm currently considering fluentd (but I've found some quite poor choices working with their previous gems also) - rsyslog, or writing my own service for syslog-endpoint -> elastic search. If you have a current approach that works well I'd be glad to know about it.

**Comment by ramiamar** [comment]
if it's of any help, I encountered the same issue, and while trying to debug it, 
I added a java opt for debugging jruby ssl:

`LS_JAVA_OPTS="-Djruby.openssl.debug=true"`

and got this "Inbound closed before receiving close_notify" exception:

```
SSLSocket.closeInbound javax.net.ssl.SSLException: Inbound closed before receiving peer's close_notify: possible truncation attack?
javax.net.ssl.SSLHandshakeException: Socket closed
	at org.jruby.ext.openssl.SSLSocket.doHandshake(SSLSocket.java:520)
	at org.jruby.ext.openssl.SSLSocket.acceptImpl(SSLSocket.java:318)
	at org.jruby.ext.openssl.SSLSocket.accept(SSLSocket.java:272)
	at org.jruby.ext.openssl.SSLSocket$INVOKER$i$0$0$accept.call(SSLSocket$INVOKER$i$0$0$accept.gen)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:134)
	at org.jruby.ast.CallNoArgNode.interpret(CallNoArgNode.java:60)
	at org.jruby.ast.IfNode.interpret(IfNode.java:116)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.ast.BlockNode.interpret(BlockNode.java:71)
	at org.jruby.ast.RescueNode.executeBody(RescueNode.java:221)
	at org.jruby.ast.RescueNode.interpret(RescueNode.java:116)
	at org.jruby.ast.BeginNode.interpret(BeginNode.java:83)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.ast.BlockNode.interpret(BlockNode.java:71)
	at org.jruby.evaluator.ASTInterpreter.INTERPRET_METHOD(ASTInterpreter.java:74)
	at org.jruby.internal.runtime.methods.InterpretedMethod.call(InterpretedMethod.java:139)
	at org.jruby.internal.runtime.methods.DefaultMethod.call(DefaultMethod.java:187)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:134)
	at org.jruby.ast.CallNoArgNode.interpret(CallNoArgNode.java:60)
	at org.jruby.ast.FCallOneArgNode.interpret(FCallOneArgNode.java:36)
	at org.jruby.ast.LocalAsgnNode.interpret(LocalAsgnNode.java:123)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.ast.BlockNode.interpret(BlockNode.java:71)
	at org.jruby.ast.RescueNode.executeBody(RescueNode.java:221)
	at org.jruby.ast.RescueNode.interpret(RescueNode.java:116)
	at org.jruby.ast.BeginNode.interpret(BeginNode.java:83)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.ast.WhileNode.interpret(WhileNode.java:131)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.ast.EnsureNode.interpret(EnsureNode.java:96)
	at org.jruby.evaluator.ASTInterpreter.INTERPRET_METHOD(ASTInterpreter.java:74)
	at org.jruby.internal.runtime.methods.InterpretedMethod.call(InterpretedMethod.java:182)
	at org.jruby.internal.runtime.methods.DefaultMethod.call(DefaultMethod.java:203)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:326)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:170)
	at org.jruby.ast.FCallOneArgNode.interpret(FCallOneArgNode.java:36)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.ast.IfNode.interpret(IfNode.java:116)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.evaluator.ASTInterpreter.INTERPRET_METHOD(ASTInterpreter.java:74)
	at org.jruby.internal.runtime.methods.InterpretedMethod.call(InterpretedMethod.java:182)
	at org.jruby.internal.runtime.methods.DefaultMethod.call(DefaultMethod.java:203)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:326)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:170)
	at org.jruby.ast.CallOneArgNode.interpret(CallOneArgNode.java:57)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.ast.RescueNode.executeBody(RescueNode.java:221)
	at org.jruby.ast.RescueNode.interpret(RescueNode.java:116)
	at org.jruby.ast.EnsureNode.interpret(EnsureNode.java:96)
	at org.jruby.ast.BeginNode.interpret(BeginNode.java:83)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.ast.BlockNode.interpret(BlockNode.java:71)
	at org.jruby.evaluator.ASTInterpreter.INTERPRET_METHOD(ASTInterpreter.java:74)
	at org.jruby.internal.runtime.methods.InterpretedMethod.call(InterpretedMethod.java:182)
	at org.jruby.internal.runtime.methods.DefaultMethod.call(DefaultMethod.java:203)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:168)
	at org.jruby.ast.FCallOneArgNode.interpret(FCallOneArgNode.java:36)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.evaluator.ASTInterpreter.INTERPRET_BLOCK(ASTInterpreter.java:112)
	at org.jruby.runtime.Interpreted19Block.evalBlockBody(Interpreted19Block.java:206)
	at org.jruby.runtime.Interpreted19Block.yield(Interpreted19Block.java:194)
	at org.jruby.runtime.Interpreted19Block.call(Interpreted19Block.java:125)
	at org.jruby.runtime.Block.call(Block.java:101)
	at org.jruby.RubyProc.call(RubyProc.java:300)
	at org.jruby.RubyProc.call(RubyProc.java:230)
	at org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:99)
	at java.lang.Thread.run(Thread.java:745)
{:timestamp=>"2016-01-16T16:01:02.047000+0000", :message=>"SSL Error", :exception=>#<OpenSSL::SSL::SSLError: Socket closed>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:272:in `accept'", "/home/logstash-2.1.1/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.14-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/home/logstash-2.1.1/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.1/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/home/logstash-2.1.1/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.1/lib/logstash/inputs/tcp.rb:84:in `run'", "/home/logstash-2.1.1/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/home/logstash-2.1.1/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error, :file=>"logstash/inputs/tcp.rb", :line=>"117", :method=>"run_server"}
IOError: closed stream
                  peeraddr at org/jruby/ext/socket/RubyIPSocket.java:95
                  peeraddr at org/jruby/ext/openssl/SSL.java:226
                      peer at /home/logstash-2.1.1/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/util/socket_peer.rb:5
             handle_socket at /home/logstash-2.1.1/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.1/lib/logstash/inputs/tcp.rb:171
  server_connection_thread at /home/logstash-2.1.1/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.1/lib/logstash/inputs/tcp.rb:144
```

**Comment by jordansissel** [comment]
Reproduced --- 

Connected with a good tls client, then telnetted in and just closed the socket w/ no data sent.

```
Accepted connection {:client=>"127.0.0.1:35936", :server=>"0.0.0.0:12345", :level=>:debug, :file=>"logstash/inputs/tcp.rb", :line=>"143", :method=>"server_connection_thread"}
config LogStash::Codecs::Line/@charset = "UTF-8" {:level=>:debug, :file=>"logstash/config/mixin.rb", :line=>"122", :method=>"config_init"}
SSL Error {:exception=>#<OpenSSL::SSL::SSLError: Unrecognized SSL message, plaintext connection?>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:262:in `accept'", "/home/jls/build/logstash-2.1.1/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.12-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/home/jls/projects/logstash-input-tcp/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/home/jls/projects/logstash-input-tcp/lib/logstash/inputs/tcp.rb:84:in `run'", "/home/jls/build/logstash-2.1.1/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:206:in `inputworker'", "/home/jls/build/logstash-2.1.1/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:199:in `start_input'"], :level=>:error, :file=>"logstash/inputs/tcp.rb", :line=>"117", :method=>"run_server"}
IOError: closed stream
                  peeraddr at org/jruby/ext/socket/RubyIPSocket.java:95
                  peeraddr at org/jruby/ext/openssl/SSL.java:209
                      peer at /home/jls/build/logstash-2.1.1/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/util/socket_peer.rb:5
             handle_socket at /home/jls/projects/logstash-input-tcp/lib/logstash/inputs/tcp.rb:171
  server_connection_thread at /home/jls/projects/logstash-input-tcp/lib/logstash/inputs/tcp.rb:144
```

**Comment by adamjt** [comment]
@jordansissel Awesome! Thanks for jumping on this.

**Comment by jordansissel** [comment]
I am working on this, but at this time have no solution. I am starting my work by writing some stronger test cases for TLS support in this plugin so that I don't break anything as a part of whatever fix comes.

**Comment by csparpa** [comment]
Thanks @jordansissel I also experience the same issue

**Comment by jordansissel** [comment]
https://github.com/logstash-plugins/logstash-input-tcp/pull/43 I think should solve this.

**Comment by adamjt** [comment]
@jordansissel #43 has fixed half of my problem - SSL errors no longer crash the logstash process, which is awesome! For any clients transmitting logs with rsyslog-gnutls and using the libgnutls26 2.12.23-12ubuntu2.4 library, all messages are dropped. This error appears in the logstash.log file:

```
{:timestamp=>"2016-02-04T00:41:21.593000+0000", :message=>"SSL Error", :exception=>#<OpenSSL::SSL::SSLError: Socket closed>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:272:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/jruby-openssl-0.9.13-java/lib/jopenssl19/openssl/ssl-internal.rb:106:in `accept'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.1/lib/logstash/inputs/tcp.rb:112:in `run_server'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-3.0.1/lib/logstash/inputs/tcp.rb:84:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/pipeline.rb:316:in `inputworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.0-java/lib/logstash/pipeline.rb:310:in `start_input'"], :level=>:error}
```

**Comment by jordansissel** [comment]
@adamjt I recommend adding some debugging to your ssl client or sniff the handshake with something like wireshark - it sounds like your client (rsyslog?) closed the connection before the handshake completed. I have no experience with rsyslog and tls, but I would recommend you check the rsyslog debug logs (if possible?) and see what it says about the problem. My assessment right now is that rsyslog is intentionally closing the connection before the handshake completes -- with no information, I would guess that it is rejecting something about the server (TLS version, SSL certificate, available ciphers, etc).

**Comment by jordansissel** [comment]
#43 has been merged and released as plugin version 3.0.2. I'm going to close this issue since I think it's resolved.
